lib/interfaces/include/BMSDriverGroup.tpp

```diff
@@ -0,0 +1,612 @@
+{
+    constexpr size_t num_chips_on_chip_select = num_chips / num_chip_selects;
+    std::array<uint8_t, 24 * num_chips_on_chip_select> combined_cv_1_to_12;
+    for (size_t chip = 0; chip < num_chips_on_chip_select; chip++)
+    {
+        size_t result_iterator = 24 * chip;
+        size_t param_iterator = 8 * chip;
+        std::copy(cv_1_to_3.begin() + param_iterator, cv_1_to_3.begin() + param_iterator + 6, combined_cv_1_to_12.begin() + result_iterator);
+        std::copy(cv_4_to_6.begin() + param_iterator, cv_4_to_6.begin() + param_iterator + 6, combined_cv_1_to_12.begin() + result_iterator + 6);
+        std::copy(cv_7_to_9.begin() + param_iterator, cv_7_to_9.begin() + param_iterator + 6, combined_cv_1_to_12.begin() + result_iterator + 12);
+        std::copy(cv_10_to_12.begin() + param_iterator, cv_10_to_12.begin() + param_iterator + 6, combined_cv_1_to_12.begin() + result_iterator + 18);
+    }
+    return combined_cv_1_to_12;
+}
+
+template <size_t num_chips, size_t num_chip_selects>
+std::array<uint8_t, 10 * (num_chips / num_chip_selects)> BMSDriverGroup<num_chips, num_chip_selects>::_package_auxillary_data(const std::array<uint8_t, 8 * (num_chips / num_chip_selects)> &aux_1_to_3,
+                                                                                                                              const std::array<uint8_t, 8 * (num_chips / num_chip_selects)> &aux_4_to_6)
+{
+    constexpr size_t num_chips_on_chip_select = num_chips / num_chip_selects;
+    std::array<uint8_t, 10 * num_chips_on_chip_select> combined_aux_1_to_5;
+    for (size_t chip = 0; chip < num_chips_on_chip_select; chip++)
+    {
+        size_t result_iterator = 10 * chip;
+        size_t param_iterator = 8 * chip;
+        std::copy(aux_1_to_3.begin() + param_iterator, aux_1_to_3.begin() + param_iterator + 6, combined_aux_1_to_5.begin() + result_iterator);
+        std::copy(aux_4_to_6.begin() + param_iterator, aux_4_to_6.begin() + param_iterator + 4, combined_aux_1_to_5.begin() + result_iterator + 6);
+    }
+    return combined_aux_1_to_5;
+}
+
+template <size_t num_chips, size_t num_chip_selects>
+void BMSDriverGroup<num_chips, num_chip_selects>::set_addresses(std::array<int, num_chips> addr)
+{
+    if (addr.size() != num_chips)
+    {
+        Serial.println("Size of set_address parameter is invalid / != num_chips");
+        delay(100000);
+    }
+    std::copy(addr.data(), addr.data() + num_chips, _address.data());
+}
+
+template <size_t num_chips, size_t num_chip_selects>
+void BMSDriverGroup<num_chips, num_chip_selects>::set_chip_select_per_chip(std::array<int, num_chips> chip_selects)
+{
+    if (chip_selects.size() != num_chips)
+    {
+        Serial.println("Size of set_address parameter is invalid / != num_chips");
+        delay(100000);
+    }
+    std::copy(chip_selects.data(), chip_selects.data() + num_chips, _chip_select_per_chip.data());
```

Comment:
same as the setter for the addresses, dont delay and dont still copy as well. also switch to using a static assert here to fail during compilation instead of runtime for checking the array size:

https://en.cppreference.com/w/cpp/language/static_assert